lib/Wikia/src/Service/User/Permissions/PermissionsServiceImpl.php

```diff
@@ -0,0 +1,391 @@
+<?php
+
+namespace Wikia\Service\User\Permissions;
+
+class PermissionsServiceImpl implements PermissionsService {
+
+	/** @var  PermissionsConfiguration */
+	private $permissionsConfiguration;
+
+	/** @var string[string] - key is user id */
+	private $localExplicitUserGroups = [];
+
+	/** @var string[string] - key is the user id */
+	private $globalExplicitUserGroups = [];
+
+	/** @var string[string] - key is the user id */
+	private $implicitUserGroups = [];
+
+	/** @var string[string] - key is user id */
+	private $userPermissions = [];
+
+	public function __construct() {
+		$this->permissionsConfiguration = new PermissionsConfigurationImpl();
+	}
+
+	public function getConfiguration() {
+		return $this->permissionsConfiguration;
+	}
+
+	public function getExplicitGroups(\User $user ) {
+		return array_unique( array_merge (
```

Comment:
coming back to "calculated every time", for sure `array_unique()` is O(n), I didn't check if that's somewhere called (directly or indirectly) in a loop but if yes that smells like a pure waste of cpu for me, anything we can do about it?